src/main/scala/scala/tools/partest/nest/StreamCapture.scala

```diff
@@ -50,4 +51,42 @@ object StreamCapture {
     }
     Captured(stdoutFn(), stderrFn(), result)
   }
+
+  def withExtraProperties[A](extra: Map[String, String])(action: => A): A = {
+    val saved = System.getProperties()
+    val modified = new java.util.Properties()
+    saved.stringPropertyNames().forEach((k) => modified.setProperty(k, saved.getProperty(k)))
+    extra.foreach { case (k, v) => modified.setProperty(k, v) }
+    System.setProperties(modified)
+    try {
+      action
+    } finally {
+      System.setProperties(saved)
+    }
+  }
+
+  // TODO merge with code above
+  def capturingOutErr[A](output: OutputStream)(f: => A): A = {
+    import java.io._
+    val savedOut = System.out
+    val savedErr = System.err
+    try {
+      val charset = Charset.defaultCharset()
```

Comment:
(Just wondering: how do we know `System.out` was set to a stream with the encoding returned by `Charset.defaultCharset()`? Does it matter? It doesn't look like we can recover the info anyway.)

Reply:
At this point `System.out` will be the default, as the rest of partest hasn't messed with it.

But looking back at this code, I now realise that one other place in the partest is mutating `System.{out,err}`:

```
pushTranscript(args mkString " ")
    val captured = StreamCapture(runCommand(args, logFile))
    if (captured.result) genPass() else {
      logFile appendAll captured.stderr
      genFail("java compilation failed")
    }
```

So we have a potential race condition if a Java compilation step happens concurrently with a non-forked test execution step.

That wrapping of Javac looks to be a no-op, we fork a process to start javac, so we should just rely on the `ProcessLogger` to redirect its output to a file. I'll add an assertion to that effect and test things out, and hopefully I'll be able to remove that use of `StreamCapture`.

Reply:
Yep, we still get Javac compilation errors under `--verbose` without that `StreamCapture`. Pushed that change.